### PR TITLE
Make Transport Shard Bulk Action Async

### PR DESCRIPTION
**Pull request by original-brownbear**
This is a dependency of #39504 

Motivation: 
By refactoring `TransportShardBulkAction#shardOperationOnPrimary` to async, we enable using `DeterministicTaskQueue` based tests to run indexing operations. This was previously impossible since we were blocking on the `write` thread until the `update` thread finished the mapping update.
With this change, the mapping update will trigger a new task in the `write` queue instead. 
This change significantly enhances the amount of coverage we get from `SnapshotResiliencyTests` (and other potential future tests) when it comes to tracking down concurrency issues with distributed state machines.

The logical change is effectively all in `TransportShardBulkAction`, the rest of the changes is then simply mechanically moving the caller code and tests to being async and passing the `ActionListener` down.

Since the move to async would've added more parameters to the `private static` steps in this logic, I decided to inline and dry up (between delete and update) the logic as much as I could instead of passing the listener + wait-consumer down through all of them.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: We are moving from the `update` to the `write` thread here after a mapping update updates the cluster state. I'm a little unsure about this tbh. Is it a problem that when we're under pressure and could get a bulk rejection that the mapping update could have happened but the write then got rejected?
I assumed not, as the same kind of inconsistency can happen if the ES process dies after the mapping update in the current implementation, but thought I should point that out (obviously this is a much more likely spot now when bulk rejections happen).

File: server/src/main/java/org/elasticsearch/cluster/action/index/MappingUpdatedAction.java
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/action/index/MappingUpdatedAction.java]: This is a little strange admittedly, but I needed part of the exception unwrapping logic from `org.elasticsearch.common.util.concurrent.FutureUtils#get(java.util.concurrent.Future<T>, long, java.util.concurrent.TimeUnit)` to get the same exception rendering as we currently have with the blocking call so I copied it. Maybe it would make sense to provide a version of `execute` that does this unwrapping automatically to make the blocking and non-blocking calls behave the same way on failures eventually?

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/packaging-sample (vagrant timed out)

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/packaging-sample (vagrant timed out again)

**Comment by original-brownbear** [comment]
@ywelsch @dnhatn any chance one of you could take a look here? This one isn't super urgent but it's blocking further progress on fixing snapshot repository stability since it holds up #39504 so it would be great if we could continue here next week or so :)

File: server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: The main reason an earlier refactoring of this class did not move this bit to async code was that this part here is problematic. We might have partially executed the batch and generated sequence numbers for it. Failing the request here now is not an option as it will otherwise have the replica go out of sync with the primary, lead to gaps in the history on the replica and block local and global checkpoint advancement, given that the partially executed request won't be replicated.
There are two options to explore: 1) forcing this onto the write executor or 2) marking the current and all subsequent bulk items as failed, and then replicate the request normally to the replicas. I would be interested in @bleskes's thoughts on this.

Also this will require extensive testing given the brittle nature of it.

Comment (bleskes) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: Indeed we avoided doing this in order to not have to deal with rejections / not to have a way bypass queue sizes (which will happen with force execution) at the time. I think the added value of the extra testing coverage merits exploring our options here. I'm inclined to go with the second suggestion (mark the rest of the operations as failed on rejection). I'm not too worried about having committed the mapping changes but ending up not using them. I think that just an edge case and as Armin mentioned - it can happen anyway.

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: @ywelsch @bleskes thanks for taking a look. I implemented option 2 real quick in https://github.com/elastic/elasticsearch/pull/39793/commits/f3b59c2730edb7cf9035e4339673e22eee258b59 (on bulk reject just loop through all remaining requests and fail them). If this looks ok, I'd look into adding test coverage for the scenario mentioned above?

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/2 (Vbox failed to come up)

File: server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: I think I was also able to write a valid test (at least I reproduced the issue that @ywelsch mentioned) in https://github.com/elastic/elasticsearch/pull/39793/commits/a3c8cac75e1340e7eca97f91efa57cef0fde709c. Take a look when you have a sec :)

File: server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: instead of having this method return a boolean and internally spawn off the mapping update, I wonder if it would be better to leave that to the caller, which will have to check after running this method whether `context.isInitial()` or `context.isMappingUpdate()` and then do the right thing. This would hopefully reduce the complexity of this method a bit, and not require an ActionListener on it.

**Review by ywelsch — CHANGES_REQUESTED** [review]
I've left some comments, the main one is about the structure in TransportShardBulkAction. I would also like @dnhatn to have a look here as well, given the complexity of the refactoring.

File: server/src/test/java/org/elasticsearch/action/bulk/BulkRejectionIT.java
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/action/bulk/BulkRejectionIT.java]: ~5s on a Ryzen 2700x

File: server/src/test/java/org/elasticsearch/action/bulk/BulkRejectionIT.java
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/action/bulk/BulkRejectionIT.java]: I just wanted to make it as likely as possible for both requests to execute in parallel so I just figured I'd save a `client()` call, it doesn't really matter though, I'll revert that :)

File: server/src/test/java/org/elasticsearch/action/bulk/BulkRejectionIT.java
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/action/bulk/BulkRejectionIT.java]: This was the only way I could slow things down enough to actually get bulk rejections when returning from a dynamic mapping update.
I found no other way of forcing this in an integration test. This would be much nicer to reproduce using something like the `DeterministicTaskQueue` but unfortunately, we don't have the infrastructure to simulate rejections there (yet?).

File: server/src/test/java/org/elasticsearch/index/seqno/RetentionLeaseBackgroundSyncActionTests.java
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/index/seqno/RetentionLeaseBackgroundSyncActionTests.java]: `WriteStateException` is checked and I directly invoked the callback from `onResponse` in the wrapped listener. I'll clean this up by using `wrap` sec :)

File: test/framework/src/main/java/org/elasticsearch/action/support/ActionTestUtils.java
Comment (original-brownbear) [reply test/framework/src/main/java/org/elasticsearch/action/support/ActionTestUtils.java]: Jup did that now :)

File: server/src/test/java/org/elasticsearch/action/bulk/BulkRejectionIT.java
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/action/bulk/BulkRejectionIT.java]: see below, just a dirty hack to enforce parallelism

File: server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: Yea that sounds better I think, I'll try that now :)

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: I gave up on this :( I did my best to at least shorten the method a little by drying up exception handling in c5edc6e74ef759070684a8a65413e107451e8f51 but it's pretty tricky to do the mapping update outside the method because its failure handling needs access to at least `updateResult` and it seems that any solution and the mapping update calll requires the engine result `result` as well, so those two would have to be set on the context somehow or so to be able to handle the mapping update outside the method. That seems like it would just make things worse in terms of readability?

**Comment by original-brownbear** [comment]
@ywelsch all points addressed :)

File: server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: Maybe we should just reduce the amount of abstraction here (it seems we only pass ` MappingUpdatePerformer mappingUpdater, Consumer<ActionListener<Void>> waitForMappingUpdate` to get testability) and handle the tests that rely on this differently to simplify things instead?

File: server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: @ywelsch I have a question here: shouldn't we handle this case (general exception in the whole body of this method, in the real world probably only https://github.com/elastic/elasticsearch/pull/39793/files#diff-720a796f6beda1dfa6af60b45ffe1010R238) the same way we handle the rejection when coming back from a mapping update and fail all subsequent unprocessed items?

**Review by ywelsch — APPROVED** [review]
@dnhatn should review this too before merging.

File: server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java]: alright, cleaned up the signature :)

**Comment by dnhatn** [comment]
@original-brownbear I am on it today.

**Comment by original-brownbear** [comment]
@ywelsch thanks! 

re https://github.com/elastic/elasticsearch/pull/39793#discussion_r272555227 (can't normally comment there for some reason GH won't let me),

I pushed https://github.com/elastic/elasticsearch/pull/39793/commits/7d6fd4aaa649739a3c6d0afa8e496e7dbc862988 (and then https://github.com/elastic/elasticsearch/pull/39793/commits/9e843857f2431615dc1bbadedc4d3eb292b2c973 to clean up one dumb change :)) to assert `false` on uncaught exceptions and also save one level of indentation as a nice side effect.

File: server/src/main/java/org/elasticsearch/action/admin/indices/close/TransportVerifyShardBeforeCloseAction.java
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/action/admin/indices/close/TransportVerifyShardBeforeCloseAction.java]: Since most of write actions execute its primary action synchronously, maybe introduce an async method which uses the result from the sync version by default (see https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/action/support/single/shard/TransportSingleShardAction.java#L106-L120).

**Review by dnhatn — APPROVED** [review]
LGTM. Thanks @original-brownbear.

I left an optional suggestion.

File: server/src/main/java/org/elasticsearch/action/admin/indices/close/TransportVerifyShardBeforeCloseAction.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/admin/indices/close/TransportVerifyShardBeforeCloseAction.java]: I tried doing this here, but ran into a bit of a mess with the slightly strange way we handle the overrides of this method where we constrain the type to `WritePrimaryResult` so I decided to not do it here to unblock another thing.
I like the idea though! => I'll open a PR that cleans this up (very) shortly :)

Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/admin/indices/close/TransportVerifyShardBeforeCloseAction.java]: No worry. It's optional. Thanks for giving it a try.

**Pull request merged**
